### PR TITLE
add CrossProduct Layer

### PR DESCRIPTION
**Pull request by sperlingxx**
## What changes were proposed in this pull request?

Add _CrossProduct_ layer 

## How was this patch tested?

unit test

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/CrossProduct.scala [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+    val (_, batch, _) = getShape(input[Tensor[T]](1))
+    output.resize(batch, len * (len - 1) / 2)
+
+    if (embeddingSize > 0) {
+      (1 to len).foreach(i => checkEmbeddingSize(input(i)))
+    }
+
+    var cc = 1
+    for( i <- 1 until len; j <- i + 1 to len ) {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/CrossProduct.scala]: maybe use while is better than for here, as it's in module forward path(compute sensitive). As far as I know there're some overhead in ```for``` in scala.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/CrossProduct.scala [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+  }
+
+  protected def checkEmbeddingSize(t: Tensor[T]): Unit = {
+    val size = if (t.dim() == 1) t.size(1) else t.size(2)
+    require(embeddingSize <= 0 || embeddingSize == size,
+      s"size of input Tensor($size) not equal to embeddingSize($embeddingSize)!")
+  }
+
+  protected def dot(t1: Tensor[T], t2: Tensor[T]): Tensor[T] = {
+    var (input1, input2) = (t1, t2)
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/CrossProduct.scala]: why not use the ```tensor.dot``` method?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/CrossProduct.scala [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    }
+
+    output
+  }
+
+  override def updateGradInput(input: Table, gradOutput: Tensor[T]): Table = {
+    gradInput = T()
+
+    if (propagateBack) {
+      val (len, gout) = input.length() -> gradOutput
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/CrossProduct.scala]: any reason to write like this? Looks like it creates an extra tuple object and then calls its extract method to assign to those two variables.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/CrossProduct.scala [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+      require(gout.dim() == 2, s"invalid dim of gradOutput(${gout.dim()})!")
+
+      val outLen = len * (len - 1) / 2
+      require(gout.size(2) == outLen,
+        s"invalid colSize of gradOutput(${gout.size(2)}), it should be $outLen!")
+
+      val (dim, _, emLen) = getShape(input[Tensor[T]](1))
+
+      var cc = 1
+      for( i <- 1 until len; j <- i + 1 to len ) {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/CrossProduct.scala]: same with the one in the forward method

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/CrossProduct.scala [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+            input[Tensor[T]](i).view(1, emLen) -> input[Tensor[T]](j).view(1, emLen)
+          case 2 =>
+            input[Tensor[T]](i) -> input[Tensor[T]](j)
+        }
+
+        // get cc_th column data from total gradOut
+        val go = gout.narrow(2, cc, 1)
+
+        val jInc = Tensor[T]().resizeAs(ti).copy(ti).cmul(go)
+        if (dim == 1) jInc.resize(Array(jInc.size(2)))
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/CrossProduct.scala]: ```tensor.squeeze```?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/CrossProduct.scala [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        gradInput.get[Tensor[T]](j) match {
+          case None => gradInput.update(j, jInc)
+          case Some(v) => v.add(jInc)
+        }
+
+        val iInc = Tensor[T]().resizeAs(tj).copy(tj).cmul(go)
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/CrossProduct.scala]: ```tensor.squeeze```?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/CrossProduct.scala (hunk 2 above)
Comment (sperlingxx) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/CrossProduct.scala]: @yiheng  We do batch dot operations for dim(2) here, which Tensor.dot don't support.

**Comment by yiheng** [comment]
LGTM. Run jenkins

**Comment by yiheng** [comment]
python test failure. For each scala layer, there should be a python wrapper... 

Please add layer at: https://github.com/intel-analytics/BigDL/blob/master/pyspark/bigdl/nn/layer.py
And add layer create method at: https://github.com/intel-analytics/BigDL/blob/master/spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala

**Comment by sperlingxx** [comment]
@yiheng python wrapper added

**Comment by yiheng** [comment]
Please check the error
```
=================================== FAILURES ===================================
____________________ [doctest] bigdl.nn.layer.CrossProduct _____________________
2568     Dimensions of input' Tensors could be one or two, if two, first dimension is `batchSize`.
2569     For convenience, output is 2-dim Tensor regardless of input' dims.
2570 
2571     Table size checking and Tensor size checking will be execute before each forward,
2572     when [[numTensor]] and [[embeddingSize]] are set values greater than zero.
2573 
2574     :param numTensor (for checking)number of Tensor input Table contains, :default 0(won't check)
2575     :param embeddingSize (for checking)vector length of dot product, :default 0(won't check)
2576 
2577     >>> crossProduct = CrossProduct()
UNEXPECTED EXCEPTION: Py4JError('An error occurred while calling o24.createCrossProduct. Trace:\npy4j.Py4JException: Method createCrossProduct([class java.lang.Integer]) does not exist\n\tat py4j.reflection.ReflectionEngine.getMethod(ReflectionEngine.java:333)\n\tat py4j.reflection.ReflectionEngine.getMethod(ReflectionEngine.java:342)\n\tat py4j.Gateway.invoke(Gateway.java:252)\n\tat py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:133)\n\tat py4j.commands.CallCommand.execute(CallCommand.java:79)\n\tat py4j.GatewayConnection.run(GatewayConnection.java:207)\n\tat java.lang.Thread.run(Thread.java:745)\n\n',)
Traceback (most recent call last):

  File "/opt/work/conda/envs/py35/lib/python3.5/doctest.py", line 1321, in __run
    compileflags, 1), test.globs)

  File "<doctest bigdl.nn.layer.CrossProduct[0]>", line 1, in <module>

  File "/opt/work/jenkins/workspace/BigDL-PRVN-pythontest-Spark-1.5-py35/pyspark/bigdl/nn/layer.py", line 2584, in __init__
    super(CrossProduct, self).__init__(None, numTensor, embeddingSize)

  File "/opt/work/jenkins/workspace/BigDL-PRVN-pythontest-Spark-1.5-py35/pyspark/bigdl/nn/layer.py", line 77, in __init__
    bigdl_type, self.jvm_class_constructor(), *args)

  File "/opt/work/jenkins/workspace/BigDL-PRVN-pythontest-Spark-1.5-py35/pyspark/bigdl/util/common.py", line 575, in callBigDlFunc
    raise error

  File "/opt/work/jenkins/workspace/BigDL-PRVN-pythontest-Spark-1.5-py35/pyspark/bigdl/util/common.py", line 568, in callBigDlFunc
    result = callJavaFunc(sc, api, *args)

  File "/opt/work/jenkins/workspace/BigDL-PRVN-pythontest-Spark-1.5-py35/pyspark/bigdl/util/common.py", line 613, in callJavaFunc
    result = func(*args)

  File "/opt/work/spark-1.5.0/python/lib/py4j-0.8.2.1-src.zip/py4j/java_gateway.py", line 538, in __call__
    self.target_id, self.name)

  File "/opt/work/spark-1.5.0/python/lib/py4j-0.8.2.1-src.zip/py4j/protocol.py", line 304, in get_return_value
    format(target_id, '.', name, value))

py4j.protocol.Py4JError: An error occurred while calling o24.createCrossProduct. Trace:
py4j.Py4JException: Method createCrossProduct([class java.lang.Integer]) does not exist
	at py4j.reflection.ReflectionEngine.getMethod(ReflectionEngine.java:333)
	at py4j.reflection.ReflectionEngine.getMethod(ReflectionEngine.java:342)
	at py4j.Gateway.invoke(Gateway.java:252)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:133)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:207)
	at java.lang.Thread.run(Thread.java:745)



/opt/work/jenkins/workspace/BigDL-PRVN-pythontest-Spark-1.5-py35/pyspark/bigdl/nn/layer.py:2577: UnexpectedException
------------------------------ Captured log call -------------------------------
java_gateway.py            470 DEBUG    Command to send: c
o24
createCrossProduct
i0
e

java_gateway.py            474 DEBUG    Answer received: xspy4j.Py4JException: Method createCrossProduct([class java.lang.Integer]) does not exist\n	at py4j.reflection.ReflectionEngine.getMethod(ReflectionEngine.java:333)\n	at py4j.reflection.ReflectionEngine.getMethod(ReflectionEngine.java:342)\n	at py4j.Gateway.invoke(Gateway.java:252)\n	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:133)\n	at py4j.commands.CallCommand.execute(CallCommand.java:79)\n	at py4j.GatewayConnection.run(GatewayConnection.java:207)\n	at java.lang.Thread.run(Thread.java:745)\n
```

BTW, here're the commands to run python test
```bash
#!/bin/bash
set -e

ulimit -a

export _JAVA_OPTIONS="-XX:MaxPermSize=3G -Xmx10G"
export MAVEN_OPTS="-XX:ReservedCodeCacheSize=512m -XX:MaxPermSize=3G -Xmx10G"
export SPARK_HOME=/opt/work/spark
mvn clean package -DskipTests
pyspark/test/dev/run-all.sh
```

**Comment by sperlingxx** [comment]
@yiheng  It seems that I forgot add __bigdl_type__ in python constructor. And it has been tested this time.

**Comment by yiheng** [comment]
jenkins pass. Merge. Thanks for the contribution

**Pull request merged**